superset/assets/javascripts/dashboard/components/GridCell.jsx

```diff
@@ -0,0 +1,108 @@
+/* eslint-disable react/no-danger */
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import SliceHeader from './SliceHeader';
+import ChartContainer from '../../chart/ChartContainer';
+
+import '../../../stylesheets/dashboard.css';
+
+const propTypes = {
+  timeout: PropTypes.number,
+  datasource: PropTypes.object,
+  isLoading: PropTypes.bool,
+  isExpanded: PropTypes.bool,
+  widgetHeight: PropTypes.number,
+  widgetWidth: PropTypes.number,
+  fetchSlice: PropTypes.func,
+  removeSlice: PropTypes.func.isRequired,
+  updateSliceName: PropTypes.func,
+  toggleExpandSlice: PropTypes.func,
+  slice: PropTypes.object,
+  chartKey: PropTypes.string,
+  formData: PropTypes.object,
+  filters: PropTypes.object,
+  addFilter: PropTypes.func,
+  clearFilter: PropTypes.func,
+  removeFilter: PropTypes.func,
+  exploreChartUrl: PropTypes.string,
+  exportCSVUrl: PropTypes.string,
+};
+
+class GridCell extends React.PureComponent {
+  getDescriptionId(slice) {
+    return 'description_' + slice.slice_id;
+  }
+
+  getHeaderId(slice) {
+    return 'header_' + slice.slice_id;
+  }
+
+  width() {
+    return this.props.widgetWidth - 10;
+  }
+
+  height(slice) {
+    const widgetHeight = this.props.widgetHeight;
+    const headerId = this.getHeaderId(slice);
+    const descriptionId = this.getDescriptionId(slice);
+    const headerHeight = this.refs[headerId] ? this.refs[headerId].offsetHeight : 30;
+    let descriptionHeight = 0;
+    if (this.props.isExpanded && this.refs[descriptionId]) {
+      descriptionHeight = this.refs[descriptionId].offsetHeight + 10;
+    }
+    return widgetHeight - headerHeight - descriptionHeight;
+  }
+
+  render() {
+    const slice = this.props.slice;
+    return (
+      <div className="slice-cell" id={`${slice.slice_id}-cell`}>
+        <div ref={this.getHeaderId(slice)}>
+          <SliceHeader
+            slice={slice}
+            exploreChartUrl={this.props.exploreChartUrl}
+            exportCSVUrl={this.props.exportCSVUrl}
+            isExpanded={this.props.isExpanded}
+            removeSlice={this.props.removeSlice}
+            updateSliceName={this.props.updateSliceName}
+            toggleExpandSlice={this.props.toggleExpandSlice}
+            forceRefresh={() => this.props.fetchSlice(true)}
+          />
+        </div>
+        <div
+          className="slice_description bs-callout bs-callout-default"
+          style={this.props.isExpanded ? {} : { display: 'none' }}
+          ref={this.getDescriptionId(slice)}
+          dangerouslySetInnerHTML={{ __html: slice.description_markeddown }}
+        />
+        <div className="row chart-container">
+          <input type="hidden" value="false" />
+          <div
+            id={'token_' + slice.slice_id}
+            className={`token col-md-12 ${this.props.isLoading ? 'is-loading' : ''}`}
+          >
+            <ChartContainer
```

Comment:
I'm surprised to see this here. I was thinking that is would go `DashboardContainer -> Dashboard -> Chart`, meaning the entry point for Redux would be at the top only. I'm not quite sure on the implications of this or whether it's the desired model, maybe @williaster can chime in here.

Reply:
Generally you can go with either "only redux at the top", or "redux in every component that needs it". The downsides of the former is that every dependency of the bottom of the chain has to be passed down through every item in the chain - which forces the intervening components to have knowledge of the implementation details of the leaves.

Reply:
@ljharb in this context we have a chart component that has a related list of actions and reducers that we reuse across apps, does that inform the Redux injection pattern in any way?

Reply:
Somewhat; if the store is injected via context, then the chart component can use the same global store, which can vary across apps - it definitely means, however, that you can't statically import the store in the leaves.

Reply:
`dashboardcontainer`'s root reducer is a combined reducer with `dashboard` reducer and `chart` reducer. I want to use `dashboard` reducer only handle dashboard state, like dashboard layout, each slice title/description data, and use `chart` reducer only handle each chart rendering data. The problem bothered me is, chart is deep nested inside Dashboard component. And middle-level components (`GridLayout`, `GridCell`) has nothing more than passing down chart data.